sockeye/decoder.py

```diff
@@ -274,48 +299,30 @@ def _step(self,
         :param seq_idx: Decoder time step.
         :return: (new decoder state, updated attention state).
         """
-        # (1) RNN step
+        # (1) First RNN step. The output will be used for computing the attention scores
         # concat previous word embedding and previous hidden state
         rnn_input = mx.sym.concat(word_vec_prev, state.hidden, dim=1,
                                   name="%sconcat_target_context_t%d" % (self.prefix, seq_idx))
         # rnn_output: (batch_size, rnn_num_hidden)
         # next_layer_states: num_layers * [batch_size, rnn_num_hidden]
-        rnn_output, layer_states = self.rnn(rnn_input, state.layer_states)
+        attention_rnn_output, attention_rnn_layer_states = \
+            self.attention_rnn(rnn_input, state.layer_states[:self.attention_rnn_n_states])
 
         # (2) Attention step
-        attention_input = self.attention.make_input(seq_idx, word_vec_prev, rnn_output)
+        attention_input = self.attention.make_input(seq_idx, word_vec_prev, attention_rnn_output)
```

Comment:
same comment here: its a bit confusing to read attention_input = .... attention_rnn_output
The lowest layer rnn isn't necessarily related to attention, its just the attention that uses it.